File: src/ide/webstorm/services/config.service.ts
import { IConfigService } from '../../../shared/types';

function normalizeUrl(url: string): string {
    return url.trim().replace(/\/+$/, '');
}

export class WebStormConfigService implements IConfigService {
    private readonly CONFIG_NAMESPACE = 'point.i18n';
    private readonly LOCALE_KEY = 'locale';
    private readonly PROJECT_KEY_KEY = 'projectKey';
    private readonly SEARCH_PATH_KEY = 'searchPath';
    private readonly API_BASE_URL_KEY = 'apiBaseUrl';
    private readonly LOCALIZATION_API_BASE_URL_KEY = 'localizationApiBaseUrl';

    private getProperties(): any {
        // @ts-ignore
        return com.intellij.ide.util.PropertiesComponent.getInstance();
    }

    getLocale(): string {
        try {
            const props = this.getProperties();
            return props.getValue(`${this.CONFIG_NAMESPACE}.${this.LOCALE_KEY}`) || 'ru';
        } catch (error) {
            console.error('Error getting locale:', error);
            return 'ru';
        }
    }

    getProjectKey(): string {
        try {
            const props = this.getProperties();
            return props.getValue(`${this.CONFIG_NAMESPACE}.${this.PROJECT_KEY_KEY}`) || 'point-frontend';
        } catch (error) {
            console.error('Error getting project key:', error);
            return 'point-frontend';
        }
    }

    getSearchPath(): string {
        try {
            const props = this.getProperties();
            return props.getValue(`${this.CONFIG_NAMESPACE}.${this.SEARCH_PATH_KEY}`) || 'src';
        } catch (error) {
            console.error('Error getting search path:', error);
            return 'src';
        }
    }

    async getApiBaseUrl(): Promise<string> {
        try {
            const props = this.getProperties();
            const url = props.getValue(`${this.CONFIG_NAMESPACE}.${this.API_BASE_URL_KEY}`);
            if (!url) {
                throw new Error('i18nRemote.apiBaseUrl не настроен');
            }
            return normalizeUrl(url);
        } catch (error) {
            console.error('Error getting API base URL:', error);
            throw error;
        }
    }

    async getLocalizationApiBaseUrl(): Promise<string> {
        try {
            const props = this.getProperties();
            const url = props.getValue(`${this.CONFIG_NAMESPACE}.${this.LOCALIZATION_API_BASE_URL_KEY}`);
            if (!url) {
                throw new Error('i18nRemote.localizationApiBaseUrl не настроен');
            }
            return normalizeUrl(url);
        } catch (error) {
            console.error('Error getting localization API base URL:', error);
            throw error;
        }
    }

    async updateLocale(locale: string): Promise<void> {
        try {
            const props = this.getProperties();
            props.setValue(`${this.CONFIG_NAMESPACE}.${this.LOCALE_KEY}`, locale);
        } catch (error) {
            console.error('Error updating locale:', error);
            throw error;
        }
    }

    async updateProjectKey(projectKey: string): Promise<void> {
        try {
            const props = this.getProperties();
            props.setValue(`${this.CONFIG_NAMESPACE}.${this.PROJECT_KEY_KEY}`, projectKey);
        } catch (error) {
            console.error('Error updating project key:', error);
            throw error;
        }
    }

    async updateSearchPath(searchPath: string): Promise<void> {
        try {
            const props = this.getProperties();
            props.setValue(`${this.CONFIG_NAMESPACE}.${this.SEARCH_PATH_KEY}`, searchPath);
        } catch (error) {
            console.error('Error updating search path:', error);
            throw error;
        }
    }

    async updateApiBaseUrl(url: string): Promise<void> {
        try {
            const props = this.getProperties();
            props.setValue(`${this.CONFIG_NAMESPACE}.${this.API_BASE_URL_KEY}`, normalizeUrl(url));
        } catch (error) {
            console.error('Error updating API base URL:', error);
            throw error;
        }
    }

    async updateLocalizationApiBaseUrl(url: string): Promise<void> {
        try {
            const props = this.getProperties();
            props.setValue(`${this.CONFIG_NAMESPACE}.${this.LOCALIZATION_API_BASE_URL_KEY}`, normalizeUrl(url));
        } catch (error) {
            console.error('Error updating localization API base URL:', error);
            throw error;
        }
    }
}

File: src/ide/webstorm/src/main/java/com/point/i18n/webstorm/services/ConfigService.java
package com.point.i18n.webstorm.services;

import com.intellij.ide.util.PropertiesComponent;

/**
 * Service for configuration.
 * Java implementation using PropertiesComponent.
 */
public class ConfigService {
    private static final String CONFIG_NAMESPACE = "point.i18n";
    private static final String API_BASE_URL_KEY = CONFIG_NAMESPACE + ".apiBaseUrl";
    private static final String LOCALIZATION_API_BASE_URL_KEY = CONFIG_NAMESPACE + ".localizationApiBaseUrl";
    private static final String DEEPL_API_KEY = CONFIG_NAMESPACE + ".deepLApiKey";
    
    private final PropertiesComponent propertiesComponent;
    
    public ConfigService() {
        this.propertiesComponent = PropertiesComponent.getInstance();
    }
    
    public String getApiBaseUrl() {
        String url = propertiesComponent.getValue(API_BASE_URL_KEY);
        if (url == null || url.trim().isEmpty()) {
            throw new IllegalStateException("API Base URL is not configured. Please configure it first.");
        }
        return normalizeUrl(url);
    }
    
    public String getLocalizationApiBaseUrl() {
        String url = propertiesComponent.getValue(LOCALIZATION_API_BASE_URL_KEY);
        if (url == null || url.trim().isEmpty()) {
            throw new IllegalStateException("Localization API Base URL is not configured. Please configure it first.");
        }
        return normalizeUrl(url);
    }
    
    public String getProjectKey() {
        return propertiesComponent.getValue(CONFIG_NAMESPACE + ".projectKey", "point-frontend");
    }
    
    public String getLocale() {
        return propertiesComponent.getValue(CONFIG_NAMESPACE + ".locale", "ru");
    }
    
    public String getSearchPath() {
        return propertiesComponent.getValue(CONFIG_NAMESPACE + ".searchPath", "src");
    }
    
    public void setProjectKey(String projectKey) {
        propertiesComponent.setValue(CONFIG_NAMESPACE + ".projectKey", projectKey);
    }
    
    public void setLocale(String locale) {
        propertiesComponent.setValue(CONFIG_NAMESPACE + ".locale", locale);
    }
    
    public void setSearchPath(String searchPath) {
        propertiesComponent.setValue(CONFIG_NAMESPACE + ".searchPath", searchPath);
    }
    
    public String getDeepLApiKey() {
        return propertiesComponent.getValue(DEEPL_API_KEY);
    }
    
    public void setDeepLApiKey(String apiKey) {
        propertiesComponent.setValue(DEEPL_API_KEY, apiKey);
    }
    
    private String normalizeUrl(String url) {
        return url.trim().replaceAll("/+$", "");
    }
}

